node/bft/src/helpers/pending.rs

```diff
@@ -146,29 +160,55 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
-        // Remove the callback for the item, and process any remaining callbacks.
-        if let Some(callbacks) = self.callbacks.lock().remove(&item) {
-            if let Some(callback_value) = callback_value {
-                // Send a notification to the callback.
-                for (callback, _, _) in callbacks {
-                    callback.send(callback_value.clone()).ok();
+        // Remove the item from the pending queue and process any remaining callbacks.
+        match self.pending.write().remove(&item) {
+            Some(callbacks) => {
+                // Get the peer IPs.
+                let peer_ips = callbacks.keys().cloned().collect();
+                // Process the callbacks.
+                if let Some(callback_value) = callback_value {
+                    // Send a notification to the callback.
+                    for (callback, _, _) in callbacks.into_values().flat_map(|callbacks| callbacks.into_iter()) {
+                        callback.send(callback_value.clone()).ok();
+                    }
                 }
+                // Return the peer IPs.
+                Some(peer_ips)
             }
+            None => None,
         }
-        // Return the result.
-        result
     }
 
     /// Removes the callbacks for the specified `item` that have expired.
-    pub fn clear_expired_callbacks_for_item(&self, item: impl Into<T>) {
+    pub fn clear_expired_callbacks_for_item(&self, now: i64, item: impl Into<T>) {
+        let item = item.into();
+
+        // Acquire the pending lock.
+        let mut pending = self.pending.write();
+
         // Clear the callbacks that have expired.
-        if let Some(callbacks) = self.callbacks.lock().get_mut(&item.into()) {
-            // Fetch the current timestamp.
-            let now = OffsetDateTime::now_utc().unix_timestamp();
-            // Remove the callbacks that have expired.
-            callbacks.retain(|(_, timestamp, _)| now - *timestamp <= CALLBACK_TIMEOUT_IN_SECS);
+        if let Some(peer_map) = pending.get_mut(&item) {
+            // Iterate over each peer IP for the item and filter out expired callbacks.
+            for (_, callbacks) in peer_map.iter_mut() {
+                callbacks.retain(|(_, timestamp, _)| now - *timestamp <= CALLBACK_EXPIRATION_IN_SECS);
+            }
+
+            // Remove peer IPs that no longer have any callbacks.
+            peer_map.retain(|_, callbacks| !callbacks.is_empty());
+
+            // If there are no more remaining callbacks for the item across all peer IPs, remove the item from pending.
+            if peer_map.is_empty() {
+                pending.remove(&item);
+            }
+        }
+    }
+
+    /// Removes the callbacks for all items have that expired.
+    pub fn clear_expired_callbacks(&self) {
+        let items = self.pending.read().keys().copied().collect::<Vec<T>>();
+        let now = OffsetDateTime::now_utc().unix_timestamp();
+        for item in items.into_iter() {
```

Comment:
once this loop is done, it might make sense to call `.shrink_to_fit` on `self.pending`, as long as we don't call this method too often

Reply:
@ljedrz We call `clear_expired_callbacks` quite often - on an interval, so I'd imagine we'd need some way to determine when it'd be appropriate to call `shrink_to_fit`.